### PR TITLE
Feature/better map state retrieval

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java [hunk 1]
```diff
@@ -265,6 +271,20 @@ public ByteBuffer verifyVirtualSubstate(SubstateId substateId)
...
+		var key = new DatabaseEntry(mapKey.array());
+		var substateId = new DatabaseEntry();
+		var result = mapDatabase.get(null, key, substateId, null);
+		if (result != SUCCESS) {
+			return Optional.empty();
+		}
+
+		var substate = loadSubstate(null, SubstateId.fromBytes(substateId.getData())).orElseThrow();
+		var substateBytes = new RawSubstateBytes(substateId.getData(), substate.array());
+		return Optional.of(substateBytes);
```
Comment (siy) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:285]: Combination of `.orElseThrow()` with return value wrapped into `Optional` looks inconsistent. Perhaps method return value should be `Result`?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java [hunk 2]
```diff
@@ -732,12 +760,64 @@ private void executeStateUpdate(com.sleepycat.je.Transaction txn, REStateUpdate
...
 				var value = new DatabaseEntry(buf2.array(), buf2.position(), buf2.remaining());
 				resourceDatabase.putNoOverwrite(txn, new DatabaseEntry(addr.getBytes()), value);
+			} else if (stateUpdate.getParsed() instanceof VirtualParent) {
+				var p = (VirtualParent) stateUpdate.getParsed();
+				var typeByte = p.getData()[0];
+				if (typeByte != SubstateTypeId.UNCLAIMED_READDR.id()) {
+					var mapKey = SystemMapKey.ofValidatorDataParent(typeByte);
+					var key = new DatabaseEntry(mapKey.array());
+					var value = new DatabaseEntry(stateUpdate.getId().asBytes());
+					var result = mapDatabase.putNoOverwrite(txn, key, value);
```
Comment (siy) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:770]: Logic above repeated few times, perhaps it worth to extract it into dedicated method.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java [hunk 3]
```diff
@@ -732,12 +760,64 @@ private void executeStateUpdate(com.sleepycat.je.Transaction txn, REStateUpdate
...
+					var p = (ValidatorData) stateUpdate.getParsed();
+					var mapKey = SystemMapKey.ofValidatorData(
+						stateUpdate.typeByte(),
+						p.getValidatorKey().getCompressedBytes()
+					);
+					var key = new DatabaseEntry(mapKey.array());
+					var result = mapDatabase.delete(txn, key);
+					if (result != SUCCESS) {
+						throw new IllegalStateException();
+					}
```
Comment (siy) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:812]: Similar to above case - logic repeated few times.

File: radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/StakeTokensConstructorV3.java [hunk 4]
```diff
@@ -65,21 +64,11 @@ public void construct(StakeTokens action, TxBuilder builder) throws TxBuilderExc
...
 		if (!flag.allowsDelegation()) {
 			final REAddr owner;
-			var validator = builder.read(
-				ValidatorOwnerCopy.class,
-				p -> p.getValidatorKey().equals(action.to()),
-				Optional.of(action.to()),
-				"Could not find state"
-			);
+			var validator = builder.read(ValidatorOwnerCopy.class, action.to());
 			owner = validator.getOwner();
```
Comment (siy) [inline radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/StakeTokensConstructorV3.java:72]: The `owner` is not used until this point. Perhaps it is worth to declare it here and eliminate default initialization.

File: radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/UnstakeOwnershipConstructor.java [hunk 5]
```diff
@@ -19,22 +19,45 @@
...
-			StakeOwnership.class,
+		if (action.amount().isZero()) {
+			throw new TxBuilderException("Must transfer > 0.");
+		}
+
+		var buf = ByteBuffer.allocate(2 + ECPublicKey.COMPRESSED_BYTES + (1 + ECPublicKey.COMPRESSED_BYTES));
+		buf.put(SubstateTypeId.STAKE_OWNERSHIP.id());
+		buf.put((byte) 0);
+		buf.put(action.from().getCompressedBytes());
+		buf.put(action.accountAddr().getBytes());
```
Comment (siy) [inline radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/UnstakeOwnershipConstructor.java:44]: I see several places where this (or similar, depending on the action) logic is repeated. Perhaps it worth to create dedicated class to handle all these calculations/serializations?

File: radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/UnstakeTokensConstructorV2.java [hunk 6]
```diff
@@ -40,13 +43,27 @@ public void construct(UnstakeTokens action, TxBuilder txBuilder) throws TxBuilde
...
 			.divide(validatorStake.getAmount());
 
-		txBuilder.swapFungible(
-			StakeOwnership.class,
+		var buf = ByteBuffer.allocate(2 + ECPublicKey.COMPRESSED_BYTES + (1 + ECPublicKey.COMPRESSED_BYTES));
+		buf.put(SubstateTypeId.STAKE_OWNERSHIP.id());
+		buf.put((byte) 0);
+		buf.put(action.from().getCompressedBytes());
+		buf.put(action.accountAddr().getBytes());
+		if (buf.hasRemaining()) {
```
Comment (siy) [inline radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/UnstakeTokensConstructorV2.java:51]: Repeated code (see previous comment)

**Review by siy — APPROVED** [review]
LGTM

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java (hunk 1 above)
Comment (talekhinezh) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:285]: It's not inconsistent in that `.orElseThrow` indicates that an unexpected error occurred which can occur if storage was changed from underneath program for example. This needs to be explicitly different from an empty or present Optional.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java (hunk 2 above)
Comment (talekhinezh) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:770]: good point

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java (hunk 3 above)
Comment (talekhinezh) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:812]: good point

File: radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/StakeTokensConstructorV3.java (hunk 4 above)
Comment (talekhinezh) [reply radixdlt-engine/src/main/java/com/radixdlt/application/tokens/construction/StakeTokensConstructorV3.java:72]: good point

**Review by iamyulong — APPROVED** [review]

**Pull request merged**